### PR TITLE
haskell: option to use a stackage lts cabal.config

### PR DESCRIPTION
**Pull request by ilovezfs**
The options hash passed to install_cabal_package can now include a
:config key, which specifies the cabal.config file that should be used.

If :config is specified, the corresponding file will be copied to the
cabal sandbox.  If no config is specified, the standard dependency
resolution is used.

HOMEBREW_CABAL_CONFIG_PATH = HOMEBREW_LIBRARY_PATH/language unless
overidden in the environment.

:config => "lts"
HOMEBREW_CABAL_CONFIG_PATH/cabal.config

:config => "lts-x.y"
HOMEBREW_CABAL_CONFIG_PATH/stackage-lts-x.y-cabal.config

The config files can be downloaded from
https://www.stackage.org/lts-x.y/cabal.config

HOMEBREW_CABAL_CONFIG_PATH/cabal.config should be a symlink to the
latest HOMEBREW_CABAL_CONFIG_PATH/stackage-lts-x.y-cabal.config, which
should be kept up to date with the latest stackage lts once it is
verified that all formulae using "lts" build against the latest, or are
transitioned to a specific "lts-x.y" because they don't build yet
against the latest.

Thanks to Zhiming Wang for suggesting this as the simplest approach.

### CODE REVIEW

File: Library/Homebrew/config.rb [hunk 1]
```diff
@@ -50,7 +50,11 @@ def mkpath
 HOMEBREW_TEMP = Pathname.new(ENV.fetch("HOMEBREW_TEMP", "/tmp"))
 
 unless defined? HOMEBREW_LIBRARY_PATH
-  HOMEBREW_LIBRARY_PATH = Pathname.new(__FILE__).realpath.parent.join("Homebrew")
+  HOMEBREW_LIBRARY_PATH = Pathname.new(__FILE__).realpath.parent
+end
+
+unless defined? HOMEBREW_CABAL_CONFIG_PATH
```
Comment (DomT4) [inline Library/Homebrew/config.rb]: IMO this is better placed in `utils` if we include it. `config` is almost the exclusive province of functions and elements that influence how `brew` _itself_ behaves.

Comment (ilovezfs) [inline Library/Homebrew/config.rb]: It affects what code paths you go down in Library/Homebrew/language/cabal.config but happy to move them wherever makes the most sense.

**Comment by ilovezfs** [comment]
@DomT4 Any idea why the first round of builds were failing? Looked like it decided to just stop while working on ==> cabal install --jobs=5 --only-dependencies at three different places all near to each other so I'm thinking a build watchdog decided it was past its allotted time and it was sent to bed without dinner.

File: Library/Homebrew/config.rb (hunk 1 above)
Comment (DomT4) [inline Library/Homebrew/config.rb]: Is there a reason it couldn't live directly in `language/haskell.rb`?

My point about `config` was just that the variables here primarily directly impact how `brew` operates, not `brew` + a function, in this situation, `brew install xyz` where `xyz` is a Haskell-using formulae.

**Comment by DomT4** [comment]
> Any idea why the first round of builds were failing?

If that was on Travis, it'll be a timeout. Travis pretty much never completes builds for formulae that need `cabal`.

Jenkins failed with this:
```
Failed to install monad-control-1.0.0.5
Build log ( /private/tmp/git-annex20160221-17876-1p90cne/git-annex-6.20160114/.cabal-sandbox/logs/monad-control-1.0.0.5.log ):
cabal:
/private/tmp/git-annex20160221-17876-1p90cne/git-annex-6.20160114/.cabal-sandbox/logs/monad-control-1.0.0.5.log:
does not exist
```

File: Library/Homebrew/config.rb (hunk 1 above)
Comment (ilovezfs) [inline Library/Homebrew/config.rb]: @DomT4 I think the bigger question is where do you actually want the config files. Then we can worry about where/how to specify the path to them.

**Comment by ilovezfs** [comment]
For what it's worth, this is what I see here:
https://gist.github.com/ilovezfs/8fdf5732ae345f301ef4

**Comment by mikemcquaid** [comment]
When/why would you want to do this?

**Comment by ilovezfs** [comment]
@mikemcquaid It makes for reproducible, predictable builds even in the face of the fact that the set of upstream packages is a moving target. So, for example, the current version of git-annex that is in master was able to build just fine at the time the bottles were generated. But today you would not be able to regenerate those very same bottles without some manual interventions because the build fails. The build can succeed again once #49158 is merged, but what is produced at that point will not necessarily be identical to what was originally built for the bottles, despite the fact that the version didn't change. If an lts config is used, then building formula version xyz today always means the same thing as building formula xyz tomorrow unless brew itself makes a change.

In addition, it can reduce the total build time since there's no need for dependency resolution. The lts config is feeding cabal-install a mutually consistent and complete set of packages.

If you ask it to build git-annex-6.20160126 with :config => "lts" it will automagically fail with an error saying that there is a requirements conflict between git-annex-6.20160126 and git-annex-6.20160114, since git-annex-6.20160114 is the version hard coded into the lts config:
https://github.com/ilovezfs/homebrew/blob/b40fa50cafa13f2842de15e44a41fa5bfab7339b/Library/Homebrew/language/stackage-lts-5.3-cabal.config#L582

**Comment by ilovezfs** [comment]
Empirical example of the build time difference:
==> Summary
🍺  /usr/local/Cellar/git-annex/6.20160114: 101 files, 80.1M, built in 13 minutes 28 seconds

==> Summary
🍺  /usr/local/Cellar/git-annex/6.20160126: 101 files, 80.3M, built in 17 minutes 31 seconds

**Comment by mikemcquaid** [comment]
Cool, I understand now, thanks. I'd rather this is a formula-specific option rather than an environment variable.

**Comment by ilovezfs** [comment]
@mikemcquaid It is a formula specific option. Example in the PR is here:
https://github.com/Homebrew/homebrew/pull/49387/files#diff-54dba270edb9e5854d59fa08a34fdb38R36
```
-    install_cabal_package :using => ["alex", "happy", "c2hs"] do
+    options = { :using => ["alex", "happy", "c2hs"] }
+    options[:config] = "lts" if build.stable?
```

**Comment by zmwangx** [comment]
The problem with using LTS and stick it in `Homebrew/Language` is, of course, who is going to maintain that `cabal.config`, and the problem of not being able to update, say, `git-annex`.

A much more reasonable solution, IMHO, is to fetch Stackage's `cabal.config` as a resource in individual formulae, stick it into the cabal sandbox, and profit. We can have something in `Language/haskell.rb` to make this process more seamless (or not — we already have the technology to do that).

**Comment by zmwangx** [comment]
By the way, Stackage's nightlies work just fine as resolvers. No need to restrict ourselves to LTS.

**Comment by ilovezfs** [comment]
@zmwangx I'm sure you've got it all figured out then. Closing the PR.

**Comment by zmwangx** [comment]
@ilovezfs Actually no, I don't have any code :smiley: Although it's quite easy to make it happen within existing framework, just put a `resource` block and a `cp` in a `git-annex` formula. Also you probably want to leave this open for other's opinions. (This PR was helpful to me too; before this I more or less had the same approach in my mind, I wouldn't have realized how part of it is troublesome without seeing it in action.)

**Comment by ilovezfs** [comment]
git-annex is already fixed since https://github.com/Homebrew/homebrew/commit/309ebef1badc2567478d2716a410053394153926 is now in master.

As for resource blocks, yes that will definitely work 100%. But it's highly repetitive compared to just adding :config => "nightly" or :config => "lts" or :config => "lts-5.3"

I didn't bother with config => "nightly" because it seems to me that the build and bottling of formula x.y.z shouldn't change due to nightlies changing.

The biggest problem with lts is going to be
>and the problem of not being able to update, say, git-annex

which is going to lead to people wanting to merge `devel` blocks with the very latest, and stress the servers with twice as many builds.

So if the only benefit is the use of "nightly" as a resolver, that's already primarily fixed by https://github.com/Homebrew/homebrew/pull/49158 unless the purpose is just build-time reduction, which doesn't really justify sprinkling resource blocks all over the place.

If using nightly as resolver is the main place you see potential benefit, why not build that into haskell.rb and leave the formulae untouched? Is there a significant use case for using the standard resolver assuming the latest nightly config isn't http 404'ing?

**Comment by zmwangx** [comment]
> git-annex is already fixed since 309ebef is now in master.

Yeah seen that earlier, remember I'm on that thread too :wink: 

> As for resource blocks, yes that will definitely work 100%. But it's highly repetitive compared to just adding :config => "nightly" or :config => "lts" or :config => "lts-5.3"

I'm fully aware of the issue, and that's what I mean by "We can have something in `language/haskell.rb` to make this process more seamless". Now that's a separate problem that I honestly have no idea how to solve, unless unchecksumed fetching within `haskell.rb` is tolerated. There's more than that which I'll point out below.

> I didn't bother with config => "nightly" because it seems to me that the build and bottling of formula x.y.z shouldn't change due to nightlies changing.

That's never been a good idea.

> So if the only benefit is the use of "nightly" as a resolver, that's already primarily fixed by #49158 unless the purpose is just build-time reduction, which doesn't really justify sprinkling resource blocks all over the place.

To me the biggest advantage is neither. You seemed to have forgot reproducible build? You don't get that with the standard resolver on a moving cache, however many jumps you allow.

> If using nightly as resolver is the main place you see potential benefit, why not build that into haskell.rb and leave the formulae untouched?

I thought about that too. But there are at least three problems:

* Unchecksumed download, mentioned above;
* You have to find the right nightly for the right formula — as with LTS, the version of a package you want isn't necessarily in the nightly you downloaded, and certainly may not be in the latest; you can sort of check the date of the latest commit on a formula and download the nightly from that day, but that's painful, brittle (think revision bump, style change, etc.), and a nightmare for developing formulae.
* If a formula is eagerly updated as soon as it comes out, it may not be in nightly yet.

> assuming the latest nightly config isn't http 404'ing?

Actually, all nightlies to date are available, thankfully.

**Comment by ilovezfs** [comment]
Reproducible build could be solved by saving the complete set of packages used into the INSTALL_RECEIPT.json or a separate cabal receipt. If a nightly resolver is added, I guess you'd want to save the name of the nightly used in the receipt.

As for the lack of checksumming, do you know if `stack` deals with that in any kind of sane way?

**Comment by zmwangx** [comment]
> do you know if `stack` deals with that in any kind of sane way?

Hmm, that was kind of a brutal hunt, but I finally located how they download the build plans. https://github.com/commercialhaskell/stack/blob/9374353c10468c3788310c75bea2a0f6290f4435/src/Stack/BuildPlan.hs#L435-L467:

```haskell
-- | Load the 'BuildPlan' for the given snapshot. Will load from a local copy
-- if available, otherwise downloading from Github.
loadBuildPlan :: (MonadIO m, MonadThrow m, MonadLogger m, MonadReader env m, HasHttpManager env, HasStackRoot env)
              => SnapName
              -> m BuildPlan
loadBuildPlan name = do
    env <- ask
    let stackage = getStackRoot env
    file' <- parseRelFile $ T.unpack file
    let fp = buildPlanDir stackage </> file'
    $logDebug $ "Decoding build plan from: " <> T.pack (toFilePath fp)
    eres <- liftIO $ decodeFileEither $ toFilePath fp
    case eres of
        Right bp -> return bp
        Left e -> do
            $logDebug $ "Decoding build plan from file failed: " <> T.pack (show e)
            ensureDir (parent fp)
            req <- parseUrl $ T.unpack url
            $logSticky $ "Downloading " <> renderSnapName name <> " build plan ..."
            $logDebug $ "Downloading build plan from: " <> url
            _ <- redownload req { checkStatus = handle404 } fp
            $logStickyDone $ "Downloaded " <> renderSnapName name <> " build plan."
            liftIO (decodeFileEither $ toFilePath fp) >>= either throwM return

  where
    file = renderSnapName name <> ".yaml"
    reponame =
        case name of
            LTS _ _ -> "lts-haskell"
            Nightly _ -> "stackage-nightly"
    url = rawGithubUrl "fpco" reponame "master" file
    handle404 (Status 404 _) _ _ = Just $ SomeException $ SnapshotNotFound name
    handle404 _ _ _              = Nothing
```

So that's it: they download build plans from the master branch of either https://github.com/fpco/lts-haskell or https://github.com/fpco/stackage-nightly, without checksums or anything. However, it's reasonably safe because they're using HTTPS, and it's their repo so they know their stuff shouldn't be compromised or corrupted without notice.

Now that I think about it, checksums in Homebrew are mainly used to prevent upstream replacement of resources without notice (I stand corrected), and that shouldn't happen to stackage.org's cabal config files. Moreover, we are already running `cabal update` in `haskell.rb`, so one more unchecksumed HTTPS download certainly shouldn't hurt.

Could you please reopen this PR, and

1. Fetch build plans on the fly, thus eliminating all the controversy around environment/configuration/storage;
2. Add support for nightlies (specific nightlies, e.g. `nightly-2016-02-20`) — actually there's nothing special about nightlies, just a path segment: https://www.stackage.org/nightly-2016-02-20/cabal.config vs https://www.stackage.org/lts-5.3/cabal.config.

? I'm thinking about something along the line of

```ruby
stable do
    url "https://hackage.haskell.org/package/git-annex-6.20160211/git-annex-6.20160211.tar.gz"
    sha256 "0a7d2dab17df27c06a935c469ced47a68d3c018cbb9254c9a735b584d413fb42"
    cabal_config "nightly-2016-02-20"
end
```

**Comment by zmwangx** [comment]
Or even

```
cabal_config "https://www.stackage.org/nightly-2016-02-20/cabal.config"
```

That's just a thought, I don't know how feasible it is to add a language-specific key.

**Comment by ilovezfs** [comment]
@xu-cheng @mikemcquaid How do you feel about a new DSL method for this? Sounds reasonable to me.

**Comment by mikemcquaid** [comment]
It feels like a DSL method is overkill here as is making it custom. Presumably we just want a particular config which we use for all our Haskell packages to ensure they are rebuilt identically (or at least in a more stable fashion)? In that case it feels like, as an attribute of Cabal itself, it should be an installed `resource` for the Cabal formula.

**Comment by zmwangx** [comment]
> It feels like a DSL method is overkill here

Sure, I also thought it was overkill, just trying to find the best way to fit it in.

> Presumably we just want a particular config... it should be an installed resource for the Cabal formula.

That won't happen, because `cabal.config` isn't really a static configuration file in the common sense. You get reproducible builds from Stackage because each Stackage config, being it nightly or lts, is a consistent "snapshot", containing **exactly one version for each package**, which is exactly why dependency resolution is sped up and all dependencies are predictable (there's simply no choice). You'll immediately get the idea if you look at one of those config files, say https://www.stackage.org/lts-5.3/cabal.config:

```
constraints: abstract-deque ==0.3,
             abstract-par ==0.3.3,
             AC-Vector ==2.3.2,
             accelerate ==0.15.1.0,
             ace ==0.6,
             acid-state ==0.14.0,
             ...
```

As such, each version of each package presumably needs a different configuration, which has to be defined in individual formulae. To me, the advantage of reproducible builds outweighs the inconvenience of another resource in each formula, but of course someone else may disagree.

**Comment by ilovezfs** [comment]
@mikemcquaid Maybe just have a brew audit rule to expect a cabal.config resource for stable builds of bottled formulae so that the products are reliably reproducible. But maybe that's too rigid.

As for `a particular config which we use for all our Haskell packages`, that doesn't sound possible according to @zmwangx
>You have to find the right nightly for the right formula — as with LTS, the version of a package you want isn't necessarily in the nightly you downloaded, and certainly may not be in the latest; you can sort of check the date of the latest commit on a formula and download the nightly from that day, but that's painful, brittle (think revision bump, style change, etc.), and a nightmare for developing formulae.

So it sounds like there's no such thing as a nightly single or a single lts that can work across the board.

**Comment by ilovezfs** [comment]
I'm getting the sense that there's little to be gained by requiring these cabal.configs and the bottle revisions should just be bumped as necessary if it's expected that bottles should be buildable in between actual formula version bumps, but even that is probably overkill.

Other than possibly in the future integrating an actual installed stack prefix into brew so that we're not building the exact same haskell dependencies over and over, there's no "there there" here and the status quo is fine as is.

**Comment by ilovezfs** [comment]
The brew bottles for brew haskell formulae are a little black-boxy, and the INSTALL_RECEIPT.json are a bit sparse/incomplete for them; however, unless I'm mistaken, the bottle build server logs for any given bottle should always be available, so if a build actually must be replicated exactly as it was for some sort of debugging, that should always be possible with like two minutes of research.

**Comment by zmwangx** [comment]
> I'm getting the sense that there's little to be gained by requiring these cabal.configs and the bottle revisions should just be bumped as necessary if it's expected that bottles should be buildable in between actual formula version bumps, but even that is probably overkill.

1. My recent experience with upgrading FFmpeg to 3.0 taught me how annoying it could be when your dependents randomly fail on the CI, and there's nothing you could do about as a dependency (except for delving head deep into fixing something you've never even heard about — I feel much more sorry for Homebrew maintainers now as they have to do this all the time). So, if I could, I would do maintainers/other contributors a favor by trying my best to reduce the number of moving parts.

2. A even bigger problem is users building from source, which are necessary when installing with options. Those are most prone to failures, and when build failures happen (not due to mishandling on those users' part) they would feel hopeless, as I have experienced first-hand. So, if I could, I would do those users a favor, too. (Although it is not guaranteed that the hard-coded config could pass building with options in the first place, it is certainly better tested, and can be fixed on Homebrew's end if necessary.)

**Comment by mikemcquaid** [comment]
Ok, I think I understand now, thanks for filling me in. In that case I think a new DSL (in `Language::Haskell`) that downloads the relevant config for a package as a resource would make sense. Even better if we can somehow smartly generate the URL (or at least part of it) somewhat automatically?

**Comment by zmwangx** [comment]
Just to make it clear: I'm not necessarily promoting for a blanket treatment of all Haskell formulae. But this is certainly a good thing for a monstrosity like `git-annex`, and again, it is achievable within existing technology.

**Comment by zmwangx** [comment]
> Even better if we can somehow smartly generate the URL (or at least part of it) somewhat automatically?

I assume you would like to smartly generate `nightly-2016-02-20` in `https://www.stackage.org/nightly-2016-02-20/cabal.config` (because there's nothing smart about other static fragments)? Please excuse me, but I'm not seeing the point, because it needs to be updated for each version of a formula, plus it's not especially hard to manually update?

It's at least possible though: you just walk yourself reverse-chronologically (or use binary search, or any other search strategy) within the nightlies until you find the one with the desired version of the package in question. Not very efficient though.

**Comment by ilovezfs** [comment]
@mikemcquaid I don't think this is needed at all, as the cabal-install dependency resolver is smart enough to handle this the vast majority of the time, and I think users building haskell formulae from source would be better off relying on a tool like stack than expecting to be able to use brew to build various option configurations all the time.

**Comment by zmwangx** [comment]
> I think users building haskell formulae from source would be better off relying on a tool like stack than expecting to be able to use brew to build various option configurations all the time.

Again, `stack` doesn't install to system locations (not with data and stuff), at least not at this point, so it doesn't work.

**Comment by zmwangx** [comment]
(Oh, and asking my old question again: why do I need to know Haskell and tooling around it to install Pandoc?)

**Comment by mikemcquaid** [comment]
I guess we can wait around and see how the increased backjumps works for us. Thanks for the great thoughts, though.

**Comment by zmwangx** [comment]
No problem, hopefully there will be fewer (or no) bug reports for git-annex and friends!

**Comment by ilovezfs** [comment]
@zmwangx Yup! It was lovely to see it turn green again in Jenkins yesterday and let's hope it stays that pretty color :)

**Pull request closed without merging**